src/utils/transform.ts

```diff
@@ -1,203 +1,142 @@
+					if (transformDependencies.indexOf(depId) === -1) transformDependencies.push(depId);
+				}
+			}
+		}
 
-					if (result == null) return previous;
+		if (result == null) return code;
+
+		if (typeof result === 'string') {
+			result = {
+				code: result,
+				ast: undefined,
+				map: undefined
+			};
+		} else if (typeof result.map === 'string') {
+			// `result.map` can only be a string if `result` isn't
+			result.map = JSON.parse(result.map);
+		}
 
-					if (typeof result === 'string') {
-						result = {
-							code: result,
-							ast: undefined,
-							map: undefined
-						};
-					} else if (typeof result.map === 'string') {
-						// `result.map` can only be a string if `result` isn't
-						result.map = JSON.parse(result.map);
-					}
+		if (Array.isArray(result.dependencies)) {
+			if (!transformDependencies) transformDependencies = [];
+			for (const dep of result.dependencies) {
+				transformDependencies.push(resolve(dirname(id), dep));
+			}
+		}
 
-					if (Array.isArray(result.dependencies)) {
-						if (!transformDependencies) transformDependencies = [];
-						for (const dep of result.dependencies) {
-							transformDependencies.push(resolve(dirname(id), dep));
-						}
-					}
+		if (result.map && typeof result.map.mappings === 'string') {
+			result.map.mappings = decode(result.map.mappings);
+		}
 
-					if (result.map && typeof result.map.mappings === 'string') {
-						result.map.mappings = decode(result.map.mappings);
-					}
+		// strict null check allows 'null' maps to not be pushed to the chain, while 'undefined' gets the missing map warning
+		if (result.map !== null) {
+			sourcemapChain.push(result.map || { missing: true, plugin: plugin.name });
+		}
 
-					// strict null check allows 'null' maps to not be pushed to the chain, while 'undefined' gets the missing map warning
-					if (result.map !== null) {
-						sourcemapChain.push(result.map || { missing: true, plugin: plugin.name });
-					}
+		ast = result.ast;
 
-					ast = result.ast;
+		return result.code;
+	}
 
-					return result.code;
-				})
-				.catch(err => {
-					if (typeof err === 'string') err = { message: err };
-					if (err.code !== 'PLUGIN_ERROR') {
+	return graph.pluginDriver
+		.hookReduceArg0<any, string>(
+			'transform',
+			[curSource, id],
+			transformReducer,
+			(pluginContext, plugin) => {
+				let emitAsset: EmitAsset;
+				({ assets, emitAsset } = createTransformEmitAsset(graph.assetsById, baseEmitAsset));
+				return {
+					...pluginContext,
+					warn(warning: RollupWarning | string, pos?: { line: number; column: number }) {
+						if (typeof warning === 'string') warning = { message: warning };
+						if (pos) augmentCodeLocation(warning, pos, curSource, id);
+						if (warning.code) warning.pluginCode = warning.code;
+						warning.id = id;
+						warning.code = 'PLUGIN_WARNING';
+						warning.plugin = plugin.name || '(anonymous plugin)';
+						warning.hook = 'transform';
+						graph.warn(warning);
+					},
```

Comment:
Again, the code for normalizing warnings and errors looks nearly identical. Maybe it could even be a central utility function to normalize warnings and errors from plugins with parameters to customize it for the respective hook and if it is a warning or error?
That way we ensure the warning/error handling is always consistent across hooks and have a single place to do changes?

Reply:
Again, I didn't want to get tooo deep into the errors right now. There's a lot that can be done here.